bundles/org.eclipse.ui.ide/src/org/eclipse/ui/views/markers/internal/ContentGeneratorDescriptor.java

```diff
@@ -185,23 +207,41 @@ public Collection<MarkerGroup> getMarkerGroups() {
 	 * @return Collection of {@link MarkerType}
 	 */
 	public Collection<MarkerType> getMarkerTypes() {
-		if (markerTypes == null) {
-			markerTypes = new HashSet<>();
-			IConfigurationElement[] markerTypeElements = configurationElement.getChildren(MarkerSupportRegistry.MARKER_TYPE_REFERENCE);
-			for (IConfigurationElement configElement : markerTypeElements) {
-				String elementName = configElement.getAttribute(MarkerSupportInternalUtilities.ATTRIBUTE_ID);
-				MarkerType[] types = MarkerTypesModel.getInstance().getType(elementName).getAllSubTypes();
-				markerTypes.addAll(Arrays.asList(types));
-				markerTypes.add(MarkerTypesModel.getInstance().getType(elementName));
-			}
-			if (markerTypes.isEmpty()) {
-				MarkerType[] types = MarkerTypesModel.getInstance().getType(IMarker.PROBLEM).getAllSubTypes();
-				markerTypes.addAll(Arrays.asList(types));
+		Set<MarkerType> markerTypes = new HashSet<>();
+		addMarkerTypesFrom(configurationElement, markerTypes);
```

Comment:
Is it possible to keep the old code here and move the new code to a method? Would make reading the diff easier.